File: src/algorithms/trees/properties/diameter-of-binary-tree/sources/DiameterOfBinaryTree_test.java
public class DiameterOfBinaryTree_test {
    static DiameterNode makeNode(int value, DiameterNode left, DiameterNode right) {
        DiameterNode node = new DiameterNode(value);
        node.left = left;
        node.right = right;
        return node;
    }

    public static void main(String[] args) {
        DiameterOfBinaryTree sol = new DiameterOfBinaryTree();

        // balanced 7-node BST: diameter is 4
        DiameterNode root1 = makeNode(4,
            makeNode(2, makeNode(1, null, null), makeNode(3, null, null)),
            makeNode(6, makeNode(5, null, null), makeNode(7, null, null)));
        assert sol.diameterOfBinaryTree(root1) == 4 : "Test 1 failed";

        // null root
        assert sol.diameterOfBinaryTree(null) == 0 : "Test 2 failed";

        // single node
        assert sol.diameterOfBinaryTree(makeNode(1, null, null)) == 0 : "Test 3 failed";

        // two-node tree
        assert sol.diameterOfBinaryTree(makeNode(1, makeNode(2, null, null), null)) == 1 : "Test 4 failed";

        // skewed tree
        DiameterNode skewed = makeNode(1, makeNode(2, makeNode(3, makeNode(4, null, null), null), null), null);
        assert sol.diameterOfBinaryTree(skewed) == 3 : "Test 5 failed";

        System.out.println("All tests passed!");
    }
}

File: src/algorithms/trees/properties/diameter-of-binary-tree/diameter-of-binary-tree_test.py
import importlib
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "sources"))
import sys

mod = importlib.import_module("diameter-of-binary-tree")
diameter_of_binary_tree = mod.diameter_of_binary_tree
TreeNode = mod.TreeNode


def make_node(value, left=None, right=None):
    node = TreeNode(value)
    node.left = left
    node.right = right
    return node


def test_balanced_7_node_bst():
    root = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))
    assert diameter_of_binary_tree(root) == 4


def test_null_root():
    assert diameter_of_binary_tree(None) == 0


def test_single_node():
    assert diameter_of_binary_tree(make_node(1)) == 0


def test_two_node_tree():
    assert diameter_of_binary_tree(make_node(1, make_node(2))) == 1


def test_skewed_tree():
    root = make_node(1, make_node(2, make_node(3, make_node(4))))
    assert diameter_of_binary_tree(root) == 3


if __name__ == "__main__":
    test_balanced_7_node_bst()
    test_null_root()
    test_single_node()
    test_two_node_tree()
    test_skewed_tree()
    print("All tests passed!")
